src/model_theory/bundled.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+
+import data.set.countable
+import model_theory.basic
+import category_theory.concrete_category.bundled
+
+/-!
+# Types and Classes of First-Order Structures
+This file defines types of bundled first-order structures and describes classes of first-order
+structures.
+
+## Main Definitions
+* The type of bundled `L`-structures is `category_theory.bundled L.Structure`, and each element has
+a canonical `L.Structure` instance.
+* `first_order.language.equiv_setoid` is an equivalence relation on bundled `L.Structure`s
+designating structures as equivalent when they are isomorphic.
+* `first_order.language.is_equiv_invariant` denotes that a class of `L.Structure`s is
+isomorphism-invariant.
+* `first_order.language.is_equiv_invariant.to_set_quotient` describes an isomorphism-invariant class
+of `L.Structure`s as a set of isomorphism classes.
+
+-/
+
+universes u v w
+
+open_locale first_order
+open set category_theory
+
+variables (L : first_order.language.{u v})
+
+@[protected] instance category_theory.bundled.Structure (M : bundled L.Structure) :
+  L.Structure M :=
+M.str
+
+namespace first_order
+namespace language
+open Structure
+
+/-! ### Isomorphism-Invariant Classes and Essential Countability -/
+
+/-- The equivalence relation on bundled `L.Structure`s indicating that they are isomorphic. -/
+instance equiv_setoid : setoid (bundled L.Structure) :=
+{ r := λ M N, nonempty (M ≃[L] N),
+  iseqv := ⟨λ M, ⟨equiv.refl L M⟩, λ M N, nonempty.map equiv.symm,
+    λ M N P, nonempty.map2 (λ MN NP, NP.comp MN)⟩ }
+
+variables {L} (K : Π (M : Type w) [L.Structure M], Prop)
+
+/-- A class `K` is isomorphism-invariant when structure isomorphic to a structure in `K` is also in
+  `K`. -/
+def is_equiv_invariant : Prop :=
+∀ (M N : Type w) [strM : L.Structure M] [strN : L.Structure N],
+  nonempty (@equiv L M N strM strN) → (@K M strM = @K N strN)
```

Comment:
The reason is I want this to apply to actual examples from the library, such as `Structure.fg`, `Structure.cg`, and some more that I haven't defined yet. The form of this determines the form of Fraisse classes, and I'm going to show that the class of all finite sets (in the empty language), the class of all finite linear orders, and the class of all finite graphs are Fraisse classes.

That said, I haven't worked out those exact examples yet. It will be easier on the abstract side to make a Fraisse class an abstract set of equivalence classes, so if you think that's worth it, I can just create an API for connecting to examples as the examples pop up.

Reply:
In concrete examples, you can still use `K (bundled.of (Structure.fg L))`.

I think it will be easier, especially in proofs. Compare:
```
  have hne : h.to_set_quotient.nonempty,
  { simp_rw [ne.def, function.funext_iff, not_forall, eq_iff_iff, iff_false, not_not] at hn,
    obtain ⟨M, strM, hM⟩ := hn,
    exact ⟨⟦⟨M, strM⟩⟧, hM⟩ },
```
with
```
  have hne : h.to_set_quotient.nonempty := nonempty.image _ _
```

Reply:
I'll try it - but I'm not sure there will be more than a few lines left of this PR - it may all fold back into the `fraisse_class` PR.